### PR TITLE
Added scan to backends

### PR DESCRIPTION
**Pull request by linxihui**
`scan` is powerful symbolic loop in `theano` and `tensorflow`. It is more general than `K.rnn`.
The implementation here adapts the well known `theano` interface for `K.scan`. 

`rnn` does not support to slices on a list of sequences and does not return a list of sequences. Plus, `rnn` is handle things like an RNN, but you may need a loop for something like, compute cumulative sum, or sliding windows, or even a new tensor operation that saves memory.

### CODE REVIEW

**Comment by farizrahman4u** [comment]
I am skeptical about this. What is it that scan can do that rnn can not? It is always better to have just one looping structure.

**Comment by ahundt** [comment]
You'll definitely need to add a detailed docstring like the one provided by [theano.scan](http://deeplearning.net/software/theano/library/scan.html).

**Comment by linxihui** [comment]
@farizrahman4u  `rnn` does not support a list of `sequences` as input, which causes a lots of efforts for me to  find a walk around way when implement `CRF` and mask support for `TimeDistributed` where I need to slices on both `inputs` and `mask`.  `rnn` has a fixed/special way to handle `mask`, which is not what you needs.

**Comment by Dref360** [comment]
What's the difference between this and `map_fn` ?

**Comment by linxihui** [comment]
@Dref360  `map_fn` apples on the each slice independently, `scan` does more,  e.g., compute the cumulative sum of a sequences. See theano.scan for detail.

**Comment by angeloskath** [comment]
There is also foldl, foldr that do reductions. And of course you can combine them to achieve anything (see #4434).

The main question is can we provide the same interface of scan (and does it make sense to) for both backends. I personally think that those well known high order functions are easier to think about and the cases that are left uncovered are so few that maybe they should be explicitly implemented for each backend.

I would also appreciate a usecase that cannot be implemented with the current backend functions or that would be simpler to implement with scan.

**Comment by ahundt** [comment]
Okay, so I decided to look at functional languages for precedent. There is definite precedent in [haskell](https://hackage.haskell.org/package/base-4.9.1.0/docs/Data-List.html#g:6) but it is much simpler and has `scanl` and `scanr`, equivalent to that boolean parameter. TensorFlow also implements [scan](https://www.tensorflow.org/api_docs/python/tf/scan).

I'd lean towards including it but I think the API should be redesigned a little more carefully rather than simply copying the theano design.

**Comment by fchollet** [comment]
Since is dropping the Theano `scan` into the backend. This is generally not what we want, since we are mostly normalizing on the TF API (with some exceptions), and TF already has a `scan`. It would be bad to have a Keras `scan` that's different from TF `scan`.

More importantly, both Theano and TensorFlow have demonstrated that `scan` is a terrible API that users don't understand and struggle to use. The `rnn` API is in large part an effort to make `scan` usable by humans. I really don't think we should include this API.
